File: tests/fixtures/grammar/constructs.py
"""Every control-flow construct the Python language reference defines.

One function per construct, so a disagreement names the construct rather
than a file total. Written from the reference's "Compound statements"
chapter and the "Boolean operations", "Conditional expressions" and
"Displays for lists, sets and dictionaries" sections of the expressions
chapter — not from recollection, which is how D112, D113 and D114 got in.

The last function is deliberately branchless: a construct that is *not* a
decision has to stay uncounted, and `with` is the one most often mistaken
for one because it has a suite.
"""


def if_statement(flag):
    if flag:
        return "yes"
    return "no"


def elif_chain(value):
    if value == 1:
        return "one"
    elif value == 2:
        return "two"
    elif value == 3:
        return "three"
    return "many"


def while_loop(limit):
    total = 0
    while total < limit:
        total += 1
    return total


def for_loop(items):
    total = 0
    for item in items:
        total += item
    return total


def try_except(source):
    try:
        return source()
    except ValueError:
        return "value"
    except KeyError:
        return "key"


def boolean_and(left, right):
    return left and right


def boolean_or(left, right):
    return left or right


def conditional_expression(flag):
    return "yes" if flag else "no"


def comprehension(items):
    return [item * 2 for item in items if item]


def match_statement(value):
    match value:
        case 1:
            return "one"
        case 2:
            return "two"
        case _:
            return "many"


def with_statement(handle):
    with handle as opened:
        return opened.read()

File: tests/fixtures/grammar/constructs.rs
// Every control-flow construct in the Rust reference that can carry a
// decision. One function per construct.

pub fn if_else(v: i32) -> i32 {
    if v > 0 {
        1
    } else if v < 0 {
        -1
    } else {
        0
    }
}

pub fn if_let(v: Option<i32>) -> i32 {
    if let Some(x) = v {
        return x;
    }
    0
}

pub fn while_loop(mut n: i32) -> i32 {
    while n > 0 {
        n -= 1;
    }
    n
}

pub fn while_let(mut items: Vec<i32>) -> i32 {
    let mut total = 0;
    while let Some(x) = items.pop() {
        total += x;
    }
    total
}

pub fn for_loop(items: &[i32]) -> i32 {
    let mut total = 0;
    for item in items {
        total += item;
    }
    total
}

pub fn loop_forever(mut n: i32) -> i32 {
    loop {
        n -= 1;
        if n < 0 {
            break;
        }
    }
    n
}

pub fn match_arms(v: i32) -> &'static str {
    match v {
        1 => "one",
        2 => "two",
        _ => "many",
    }
}

pub fn boolean_operators(a: bool, b: bool, c: bool) -> bool {
    a && b || c
}

pub fn question_mark(text: &str) -> Result<i32, std::num::ParseIntError> {
    let value = text.parse::<i32>()?;
    Ok(value + 1)
}
